### PR TITLE
CAL-79 Improve support for parsing TREs from NITFs

### PR DESCRIPTION
**Pull request by roelens8**
#### What does this PR do?
 -Adds the capability to parse TREs when adding TRE attributes as instance fields into the enum TreDescriptor.java

-Parses TREs from both the nitf header and image segment.

#### Who is reviewing it?
@dcruver @jaymcnallie

#### How should this be tested?
Ingest a NITF with a TRE and verify that the metacard was populated with the fields from the NITF and its TREs.

#### What are the relevant tickets?
CAL-79
#### Checklist:
- [ ] Documentation Updated
- [X] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

**Comment by bdeining** [comment]
Can we add the CAL ticket number to the PR title?

**Comment by dcruver** [comment]
This branch was created from the CAL-49 branch.  That one should be merged first.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/SegmentHandler.java [hunk 1]
```diff
@@ -32,6 +37,19 @@
...
+                .getTREs();
+        tres.stream()
+                .forEach(tre -> Arrays.stream(TreDescriptor.values())
+                        .filter(treDescriptor -> tre.getName()
+                                .trim()
+                                .equals(treDescriptor.name()))
+                        .findFirst()
+                        .ifPresent(first -> handleSegmentHeader(metacard, tre, first.getValues())));
+    }
+
```
Comment (dcruver) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/SegmentHandler.java]: I think this can be simplified.  https://gist.github.com/dcruver/454db70c8052d03f6da19dce83f83a56

Comment (roelens8) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/SegmentHandler.java]: Fixed.

**Comment by dcruver** [comment]
LGMT :+1:

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AimidbAttribute.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            Function<Tre, Serializable> accessorFunction) {
+        this.longName = longName;
+        this.shortName = shortName;
+        this.accessorFunction = accessorFunction;
+        this.attributeDescriptor = new AttributeDescriptorImpl(ATTRIBUTE_NAME_PREFIX + longName,
+                true,
+                true,
+                false,
+                true,
+                BasicTypes.STRING_TYPE);
```
Comment (jlcsmith) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AimidbAttribute.java]: Are all of these attributes be stored as a string type?  This doesn't look right to me.

Comment (roelens8) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AimidbAttribute.java]: I added a new constructor to allow storing attributes as different types and I updated the attributes to be stored as the correct type according to the specification.

**Comment by dcruver** [comment]
App built fine but failed to deploy:

Caused by: org.osgi.service.resolver.ResolutionException: Unable to resolve root: missing requirement [root] osgi.identity; osgi.identity=imaging-app; type=karaf.feature; version="[0.2.0.SNAPSHOT,0.2.0.SNAPSHOT]"; filter:="(&(osgi.identity=imaging-app)(type=karaf.feature)(version>=0.2.0.SNAPSHOT)(version<=0.2.0.SNAPSHOT))" [caused by: Unable to resolve imaging-app/0.2.0.SNAPSHOT: missing requirement [imaging-app/0.2.0.SNAPSHOT] osgi.identity; osgi.identity=imaging-transformer-nitf; type=osgi.bundle; version="[0.2.0.SNAPSHOT,0.2.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve imaging-transformer-nitf/0.2.0.SNAPSHOT: missing requirement [imaging-transformer-nitf/0.2.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=ddf.catalog.data.types)(version>=2.10.0)(!(version>=3.0.0)))"]]

**Comment by jlcsmith** [comment]
This PR is going to conflict with the NITF transformer taxonomy updates @lcrosenbu is making.  We should discuss how to ensure this capability doesn't get lost.

File: distribution/alliance/pom.xml [hunk 3]
```diff
@@ -294,27 +294,27 @@
...
-                            </artifactItems>
-                            <overWriteSnapshots>true</overWriteSnapshots>
-                        </configuration>
-                    </execution>
+                    <!--<execution>-->
+                        <!--<id>unpack-alliance-documentation</id>-->
+                        <!--<phase>generate-resources</phase>-->
+                        <!--<goals>-->
+                            <!--<goal>unpack</goal>-->
+                        <!--</goals>-->
```
Comment (bdeining) [inline distribution/alliance/pom.xml]: Why was this commented out?  Can we just remove it?

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/NitfGmtiTransformer.java [hunk 4]
```diff
@@ -78,23 +79,28 @@ public Metacard transform(NitfSegmentsFlow nitfSegmentsFlow, Metacard metacard)
...
+                    try {
+                        List<TreGroup> targets = tre.getEntry(TARGETS)
+                                .getGroups();
+
+                        targets.stream()
+                                .forEach(group -> handleSegmentHeader(metacard,
+                                        group,
+                                        IndexedMtirpbAttribute.values()));
+                    } catch (NitfFormatException e) {
+                        LOGGER.warn("Could not parse NITF target information: " + e.getMessage(),
```
Comment (bdeining) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/NitfGmtiTransformer.java]: `LOGGER.warn("Could not parse NITF target information: {}", e.getMessage(), e);`

File: distribution/alliance/pom.xml (hunk 3 above)
Comment (roelens8) [inline distribution/alliance/pom.xml]: This was commented out by mistake. It is for unpacking the alliance documentation.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/NitfGmtiTransformer.java (hunk 4 above)
Comment (roelens8) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/NitfGmtiTransformer.java]: Fixed.

**Comment by bdeining** [comment]
:+1:

**Comment by jaymcnallie** [comment]
@jlcsmith is there still a potential of merge conflicts with @lcrosenbu's ticket?

**Comment by jlcsmith** [comment]
@lcrosenbu may be able to answer that better than me.  Also, @roelens8 I'm seeing a checkstyle failure in NitfImageTransformer

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AimidbAttribute.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+import java.util.function.Function;
+
+import org.codice.imaging.nitf.core.tre.Tre;
+
+import ddf.catalog.data.AttributeDescriptor;
+import ddf.catalog.data.AttributeType;
+import ddf.catalog.data.impl.AttributeDescriptorImpl;
+import ddf.catalog.data.impl.BasicTypes;
+
+public enum AimidbAttribute implements NitfAttribute<Tre> {
```
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AimidbAttribute.java]: Do we intend on adding all the `AIMIDB` attributes to the metacard? In https://github.com/codice/alliance/pull/113, we moved to only including those values in the NITF that are mapped to the normalized taxonomy. I believe `SegmentHandler` will add all these values to the metacard.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AimidbAttribute.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+    private AttributeDescriptor attributeDescriptor;
+
+    AimidbAttribute(String longName, String shortName,
+            Function<Tre, Serializable> accessorFunction) {
+        this(longName, shortName, accessorFunction, BasicTypes.STRING_TYPE);
+    }
+
+    AimidbAttribute(final String lName, final String sName,
+            final Function<Tre, Serializable> function, AttributeType attributeType) {
```
Comment (lcrosenbu) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AimidbAttribute.java]: Modify this constructor to take either the `MetacardType` or `AttributeDescriptor` (similar to what was done above in `AcftbAttribute` rather than the `AttributeType`.

**Comment by lcrosenbu** [comment]
Recommend adding these new NITF TREs to the Confluence pages detailing the mappings to the normalized taxonomy: 
https://codice.atlassian.net/wiki/display/DDF/NITF

**Comment by lcrosenbu** [comment]
The PR that would have conflicts is https://github.com/codice/alliance/pull/113, which has already been merged.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AimidbAttribute.java (hunk 5 above)
Comment (roelens8) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AimidbAttribute.java]: Country is important, the others are are not as important, but can be included if they have mappings to the normalized taxonomy.

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AimidbAttribute.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+import org.codice.imaging.nitf.core.tre.Tre;
+
+import ddf.catalog.data.AttributeDescriptor;
+import ddf.catalog.data.MetacardType;
+import ddf.catalog.data.impl.types.LocationAttributes;
+import ddf.catalog.data.types.Location;
+
+public enum AimidbAttribute implements NitfAttribute<Tre> {
+    COUNTRY(Location.COUNTRY_CODE,
```
Comment (jlcsmith) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AimidbAttribute.java]: can this field have multiple country codes?  Would all of these be set on the metacard `location.country-code` attribute?

Comment (roelens8) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AimidbAttribute.java]: According to the spec, country is a two letter code defining the country for the reference point of the image. I do not believe it can have multiple country codes.

Comment (jlcsmith) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AimidbAttribute.java]: Ok, if its 2 letter we'll need to translate that to a 3 letter country code.  https://codice.atlassian.net/wiki/display/DDF/location.country-code

Comment (bdeining) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AimidbAttribute.java]: Can we leverage some of the code in https://github.com/codice/ddf/pull/1171?

Comment (roelens8) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AimidbAttribute.java]: Looks like we can use the ISOFormatConverter from codice/ddf#1171, to convert the country code.

Comment (dcruver) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AimidbAttribute.java]: From the spec for AIMIDB: "Country Code. Two letter code defining the country for the reference point of the image. Standard codes may be found in FIPS PUB 10-4."

The ISOFormatConverter doesn't appear to handle FIPS. We'll also need to determine how we want to handle non-standard country codes.

Comment (peterhuffer) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/common/AimidbAttribute.java]: Here is a text file from opengeocode containing a table of country codes in FIPS and corresponding ISO formats.
http://www.opengeocode.org/download/fips2iso.txt

File: catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/TestCommonTreAttributes.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import static org.junit.Assert.assertThat;
+
+import java.util.stream.Stream;
+
+import org.codice.alliance.transformer.nitf.common.AcftbAttribute;
+import org.codice.alliance.transformer.nitf.common.AimidbAttribute;
+import org.codice.imaging.nitf.core.common.NitfFormatException;
+import org.junit.Test;
+
+public class TestCommonTreAttributes {
```
Comment (bdeining) [inline catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/TestCommonTreAttributes.java]: According to testing guidelines, this test class should be named `CommonTreAttributesTest.java`

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/NitfGmtiTransformer.java [hunk 9]
```diff
@@ -42,23 +42,22 @@
...
     private static final String MTIRPB = "MTIRPB";
 
     private static final String TARGETS = "TARGETS";
 
     private static final Logger LOGGER = LoggerFactory.getLogger(NitfGmtiTransformer.class);
 
     // Handles locations that have either 6 or 7 decimal places
-    private static final String LOCATION_REGEX = "([+\\-]\\d{2}+\\.\\d{6,7}+)([+\\-]\\d{3}+\\.\\d{6,7})";
+    private static final String LOCATION_REGEX =
+            "([+\\-]\\d{2}+\\.\\d{6,7}+)([+\\-]\\d{3}+\\.\\d{6,7})";
```
Comment (bdeining) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/NitfGmtiTransformer.java]: I think we should use Pattern.compile here for performance

File: catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/NitfGmtiTransformer.java [hunk 10]
```diff
@@ -108,27 +112,31 @@ private void transformTargetLocation(Metacard metacard) {
...
-                LOGGER.debug("Setting the metacard attribute [{}, {}]", Core.LOCATION,
+                LOGGER.debug("Setting the metacard attribute [{}, {}]",
+                        Core.LOCATION,
                         geometry.toText());
                 metacard.setAttribute(new AttributeImpl(Core.LOCATION, geometry.toText()));
             }
 
         } catch (ParseException e) {
-            LOGGER.debug(e.getMessage(), e);
+            LOGGER.warn(e.getMessage(), e);
```
Comment (bdeining) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/NitfGmtiTransformer.java]: Why this change?

Comment (roelens8) [inline catalog/imaging/imaging-transformer-nitf/src/main/java/org/codice/alliance/transformer/nitf/gmti/NitfGmtiTransformer.java]: Not intentional, was from incorrectly merging.

File: catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/TestCommonTreAttributes.java (hunk 8 above)
Comment (roelens8) [inline catalog/imaging/imaging-transformer-nitf/src/test/java/org/codice/alliance/transformer/nitf/TestCommonTreAttributes.java]: Will change.

**Comment by roelens8** [comment]
Removed the AIMIDB attribute from this PR to speed up progress and prevent https://github.com/codice/alliance/pull/143/ from being held up.

I will create a new PR with the AIMIDB attribute, along with new changes to convert country codes from FIPS into ISO formats.

**Comment by dcruver** [comment]
LGTM 👍

**Comment by roelens8** [comment]
@jaymcnallie are we good to go on this?

**Comment by jaymcnallie** [comment]
👍
